File: postgres_backup.py
"""
Модуль для резервного копирования PostgreSQL баз данных
"""
import asyncio
import subprocess
import os
import gzip
from datetime import datetime
from pathlib import Path
from typing import Optional, Dict, Any
import logging

from config import settings
from cryptography.fernet import Fernet
import base64

logger = logging.getLogger(__name__)

# Ключ для шифрования паролей (в продакшене должен быть в переменных окружения)
ENCRYPTION_KEY = os.getenv("BACKUP_ENCRYPTION_KEY", settings.secret_key[:32].ljust(32, '0'))


def encrypt_password(password: str) -> str:
    """Шифрует пароль для хранения в БД"""
    try:
        # Используем секретный ключ приложения
        key_bytes = ENCRYPTION_KEY.encode()[:32].ljust(32, b'0')
        key = base64.urlsafe_b64encode(key_bytes)
        f = Fernet(key)
        return f.encrypt(password.encode()).decode()
    except Exception as e:
        logger.error(f"Error encrypting password: {e}")
        # В случае ошибки возвращаем как есть (небезопасно, но для отладки)
        return password


def decrypt_password(encrypted: str) -> str:
    """Расшифровывает пароль из БД"""
    try:
        key_bytes = ENCRYPTION_KEY.encode()[:32].ljust(32, b'0')
        key = base64.urlsafe_b64encode(key_bytes)
        f = Fernet(key)
        return f.decrypt(encrypted.encode()).decode()
    except Exception as e:
        logger.error(f"Error decrypting password: {e}")
        return ""


class PostgresBackupExecutor:
    """Класс для выполнения резервного копирования PostgreSQL"""
    
    def __init__(self, task, storage_config=None, s3_config=None):
        self.task = task
        self.storage_config = storage_config  # Новое универсальное хранилище
        self.s3_config = s3_config  # Для обратной совместимости
        self.temp_dir = Path("/tmp/postgres_backups")
        self.temp_dir.mkdir(exist_ok=True, parents=True)
    
    async def execute_backup(self) -> Dict[str, Any]:
        """Выполняет резервное копирование PostgreSQL базы данных"""
        start_time = datetime.utcnow()
        result = {
            "success": False,
            "dump_filename": None,
            "dump_size_mb": 0,
            "s3_path": None,
            "error": None
        }
        
        try:
            # Расшифровываем пароль
            password = decrypt_password(self.task.password)
            
            # Формируем имя файла дампа
            timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
            db_name_safe = self.task.database.replace("/", "_").replace("\\", "_")
            
            if self.task.backup_format == "custom":
                dump_filename = f"{db_name_safe}_{timestamp}.dump"
            elif self.task.backup_format == "plain":
                dump_filename = f"{db_name_safe}_{timestamp}.sql"
            elif self.task.backup_format == "tar":
                dump_filename = f"{db_name_safe}_{timestamp}.tar"
            else:
                dump_filename = f"{db_name_safe}_{timestamp}.sql"
            
            dump_path = self.temp_dir / dump_filename
            
            # Формируем команду pg_dump
            cmd = [
                "pg_dump",
                f"--host={self.task.host}",
                f"--port={self.task.port}",
                f"--username={self.task.username}",
                f"--dbname={self.task.database}",
                f"--format={self.task.backup_format}",
                f"--file={dump_path}",
            ]
            
            # Добавляем опции в зависимости от формата
            if self.task.backup_format == "custom":
                cmd.append(f"--compress={self.task.compression_level}")
            elif self.task.backup_format == "plain":
                cmd.append("--no-owner")
                cmd.append("--no-privileges")
            
            # Опции включения/исключения
            if not self.task.include_schema and not self.task.include_data:
                # Если оба выключены, включаем оба (по умолчанию)
                pass
            elif not self.task.include_schema:
                cmd.append("--data-only")
            elif not self.task.include_data:
                cmd.append("--schema-only")
            
            if self.task.include_roles:
                cmd.append("--roles-only")
            
            if self.task.include_tablespaces:
                cmd.append("--tablespaces")
            
            # Устанавливаем переменную окружения с паролем
            env = os.environ.copy()
            env["PGPASSWORD"] = password
            
            logger.info(f"Executing pg_dump for database {self.task.database} on {self.task.host}")
            
            # Выполняем pg_dump
            process = await asyncio.create_subprocess_exec(
                *cmd,
                stdout=asyncio.subprocess.PIPE,
                stderr=asyncio.subprocess.PIPE,
                env=env
            )
            
            stdout, stderr = await process.communicate()
            
            if process.returncode != 0:
                error_msg = stderr.decode() if stderr else "Unknown error"
                logger.error(f"pg_dump failed: {error_msg}")
                result["error"] = error_msg
                return result
            
            # Проверяем размер файла
            if dump_path.exists():
                dump_size = dump_path.stat().st_size
                result["dump_size_mb"] = dump_size / (1024 * 1024)
                result["dump_filename"] = dump_filename
                
                # Загружаем в хранилище (S3, SFTP, NFS или локальное)
                from storage_manager import StorageManager
                
                # Определяем тип хранилища и конфигурацию
                if self.storage_config:
                    # Используем новое универсальное хранилище
                    storage_type = self.storage_config.storage_type
                    config_data = self.storage_config.config_data if isinstance(self.storage_config.config_data, dict) else {}
                    remote_path = f"postgres_backups/{db_name_safe}/{dump_filename}"
                elif self.s3_config:
                    # Обратная совместимость с S3Config
                    storage_type = "s3"
                    config_data = {
                        "endpoint": self.s3_config.endpoint,
                        "access_key": self.s3_config.access_key,
                        "secret_key": self.s3_config.secret_key,
                        "bucket_name": self.s3_config.bucket_name,
                        "region": self.s3_config.region,
                        "use_ssl": self.s3_config.use_ssl
                    }
                    remote_path = f"postgres_backups/{db_name_safe}/{dump_filename}"
                else:
                    result["error"] = "No storage configuration provided"
                    return result
                
                try:
                    storage_path = await StorageManager.upload_backup(
                        storage_type,
                        config_data,
                        str(dump_path),
                        remote_path
                    )
                    result["s3_path"] = storage_path  # Сохраняем как s3_path для совместимости
                    result["success"] = True
                    logger.info(f"Backup uploaded to {storage_type}: {storage_path}")
                except Exception as e:
                    logger.error(f"Failed to upload to {storage_type}: {e}")
                    result["error"] = f"Storage upload failed: {str(e)}"
                    return result
                
                # Удаляем локальный файл после успешной загрузки
                try:
                    dump_path.unlink()
                    logger.info(f"Local dump file removed: {dump_path}")
                except Exception as e:
                    logger.warning(f"Failed to remove local dump file: {e}")
            else:
                result["error"] = "Dump file was not created"
                return result
            
        except Exception as e:
            logger.error(f"Error during PostgreSQL backup: {e}")
            result["error"] = str(e)
        
        return result
    
    async def restore_backup(self, storage_path: str, target_database: Optional[str] = None) -> Dict[str, Any]:
        """Восстанавливает базу данных из резервной копии"""
        result = {
            "success": False,
            "error": None
        }
        
        try:
            # Расшифровываем пароль
            password = decrypt_password(self.task.password)
            
            # Определяем целевую БД
            restore_db = target_database or self.task.database
            
            # Загружаем файл из хранилища
            from storage_manager import StorageManager
            
            # Определяем тип хранилища и конфигурацию
            if self.storage_config:
                # Используем новое универсальное хранилище
                storage_type = self.storage_config.storage_type
                config_data = self.storage_config.config_data if isinstance(self.storage_config.config_data, dict) else {}
            elif self.s3_config:
                # Обратная совместимость с S3Config
                storage_type = "s3"
                config_data = {
                    "endpoint": self.s3_config.endpoint,
                    "access_key": self.s3_config.access_key,
                    "secret_key": self.s3_config.secret_key,
                    "bucket_name": self.s3_config.bucket_name,
                    "region": self.s3_config.region,
                    "use_ssl": self.s3_config.use_ssl
                }
            else:
                result["error"] = "No storage configuration provided"
                return result
            
            # Извлекаем имя файла из пути хранилища
            # Формат пути: s3://bucket/path, sftp://host/path, nfs://server/path, local://path
            if "/" in storage_path:
                filename = storage_path.split("/")[-1]
            else:
                filename = Path(storage_path).name
            
            local_file = self.temp_dir / filename
            
            try:
                await StorageManager.download_backup(
                    storage_type,
                    config_data,
                    storage_path,
                    str(local_file)
                )
                logger.info(f"Downloaded backup from {storage_type}: {local_file}")
            except Exception as e:
                result["error"] = f"Failed to download from {storage_type}: {str(e)}"
                return result
            
            # Определяем формат по расширению
            file_ext = local_file.suffix.lower()
            if file_ext == ".dump" or file_ext == ".custom":
                format_type = "custom"
                restore_cmd = "pg_restore"
            elif file_ext == ".tar":
                format_type = "tar"
                restore_cmd = "pg_restore"
            else:
                format_type = "plain"
                restore_cmd = "psql"
            
            # Формируем команду восстановления
            if restore_cmd == "pg_restore":
                cmd = [
                    "pg_restore",
                    f"--host={self.task.host}",
                    f"--port={self.task.port}",
                    f"--username={self.task.username}",
                    f"--dbname={restore_db}",
                    "--clean",
                    "--if-exists",
                    str(local_file)
                ]
            else:
                # Для plain SQL используем psql
                cmd = [
                    "psql",
                    f"--host={self.task.host}",
                    f"--port={self.task.port}",
                    f"--username={self.task.username}",
                    f"--dbname={restore_db}",
                    "--file", str(local_file)
                ]
            
            env = os.environ.copy()
            env["PGPASSWORD"] = password
            
            logger.info(f"Restoring database {restore_db} from {local_file}")
            
            # Выполняем восстановление
            process = await asyncio.create_subprocess_exec(
                *cmd,
                stdout=asyncio.subprocess.PIPE,
                stderr=asyncio.subprocess.PIPE,
                env=env
            )
            
            stdout, stderr = await process.communicate()
            
            if process.returncode != 0:
                error_msg = stderr.decode() if stderr else "Unknown error"
                logger.error(f"Restore failed: {error_msg}")
                result["error"] = error_msg
            else:
                result["success"] = True
                logger.info(f"Database {restore_db} restored successfully")
            
            # Удаляем локальный файл
            try:
                local_file.unlink()
            except Exception as e:
                logger.warning(f"Failed to remove local restore file: {e}")
            
        except Exception as e:
            logger.error(f"Error during PostgreSQL restore: {e}")
            result["error"] = str(e)
        
        return result

File: agent/internal/backup/postgres_backup.go
package backup

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"mime/multipart"
	"net/http"
	"net/url"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"time"

	"backup-server-agent/internal/config"
	"backup-server-agent/internal/logger"
	"github.com/minio/minio-go/v7"
	"github.com/minio/minio-go/v7/pkg/credentials"
	"github.com/minio/minio-go/v7/pkg/credentials"
)

type PostgresBackupResult struct {
	Success      bool
	DumpPath     string
	DumpSizeMB   float64
	DumpFilename string
	StoragePath  string
	Error        string
}

// ExecutePostgresBackup выполняет резервное копирование PostgreSQL базы данных
func ExecutePostgresBackup(task config.PostgresTask, conn config.PostgresConnection, serverIP string, log *logger.Logger) (*PostgresBackupResult, error) {
	result := &PostgresBackupResult{}

	// Расшифровываем пароль (в будущем можно добавить шифрование)
	password := conn.Password

	// Формируем имя файла дампа
	timestamp := time.Now().Format("20060102_150405")
	dbNameSafe := strings.ReplaceAll(strings.ReplaceAll(task.Database, "/", "_"), "\\", "_")

	var dumpFilename string
	switch task.BackupFormat {
	case "custom":
		dumpFilename = fmt.Sprintf("%s_%s.dump", dbNameSafe, timestamp)
	case "plain":
		dumpFilename = fmt.Sprintf("%s_%s.sql", dbNameSafe, timestamp)
	case "tar":
		dumpFilename = fmt.Sprintf("%s_%s.tar", dbNameSafe, timestamp)
	default:
		dumpFilename = fmt.Sprintf("%s_%s.sql", dbNameSafe, timestamp)
	}

	// Создаем временную директорию для дампов
	tempDir := "/tmp/postgres_backups"
	if err := os.MkdirAll(tempDir, 0755); err != nil {
		result.Error = fmt.Sprintf("Failed to create temp directory: %v", err)
		return result, fmt.Errorf(result.Error)
	}

	dumpPath := filepath.Join(tempDir, dumpFilename)

	// Формируем команду pg_dump
	cmd := exec.Command(
		"pg_dump",
		fmt.Sprintf("--host=%s", conn.Host),
		fmt.Sprintf("--port=%d", conn.Port),
		fmt.Sprintf("--username=%s", conn.Username),
		fmt.Sprintf("--dbname=%s", task.Database),
		fmt.Sprintf("--format=%s", task.BackupFormat),
		fmt.Sprintf("--file=%s", dumpPath),
	)

	// Добавляем опции в зависимости от формата
	if task.BackupFormat == "custom" {
		cmd.Args = append(cmd.Args, fmt.Sprintf("--compress=%d", task.CompressionLevel))
	} else if task.BackupFormat == "plain" {
		cmd.Args = append(cmd.Args, "--no-owner", "--no-privileges")
	}

	// Опции включения/исключения
	if !task.IncludeSchema && !task.IncludeData {
		// Если оба выключены, включаем оба (по умолчанию)
	} else if !task.IncludeSchema {
		cmd.Args = append(cmd.Args, "--data-only")
	} else if !task.IncludeData {
		cmd.Args = append(cmd.Args, "--schema-only")
	}

	// Примечание: pg_dump не поддерживает флаг --roles-only
	// Роли нужно дампить отдельно через pg_dumpall --roles-only
	// Если нужно включить роли в дамп, это делается автоматически при дампе схемы
	// Для отдельного дампа ролей нужно использовать pg_dumpall --roles-only
	// Здесь мы просто пропускаем этот флаг, так как он не поддерживается в pg_dump
	// if task.IncludeRoles {
	//     cmd.Args = append(cmd.Args, "--roles-only")  // Не поддерживается в pg_dump
	// }

	// Tablespaces включены по умолчанию в pg_dump
	// Если нужно исключить tablespaces, используем --no-tablespaces
	if !task.IncludeTablespaces {
		cmd.Args = append(cmd.Args, "--no-tablespaces")
	}

	// Устанавливаем переменную окружения с паролем
	cmd.Env = append(os.Environ(), fmt.Sprintf("PGPASSWORD=%s", password))

	log.Infof("Executing pg_dump for database %s on %s:%d", task.Database, conn.Host, conn.Port)

	// Выполняем pg_dump
	output, err := cmd.CombinedOutput()
	if err != nil {
		errorMsg := string(output)
		if errorMsg == "" {
			errorMsg = err.Error()
		}
		log.Errorf("pg_dump failed: %s", errorMsg)
		result.Error = errorMsg
		return result, fmt.Errorf("pg_dump failed: %s", errorMsg)
	}

	// Проверяем размер файла
	stat, err := os.Stat(dumpPath)
	if err != nil {
		result.Error = fmt.Sprintf("Dump file was not created: %v", err)
		return result, fmt.Errorf(result.Error)
	}

	result.DumpSizeMB = float64(stat.Size()) / (1024 * 1024)
	result.DumpFilename = dumpFilename
	result.DumpPath = dumpPath

	// Загружаем в хранилище
	storagePath, err := uploadPostgresBackupToStorage(task, dumpPath, dbNameSafe, log)
	if err != nil {
		result.Error = fmt.Sprintf("Failed to upload to storage: %v", err)
		// Удаляем локальный файл даже при ошибке загрузки
		os.Remove(dumpPath)
		return result, fmt.Errorf(result.Error)
	}

	result.StoragePath = storagePath
	result.Success = true

	// Удаляем локальный файл после успешной загрузки
	if err := os.Remove(dumpPath); err != nil {
		log.Warnf("Failed to remove local dump file: %v", err)
	} else {
		log.Infof("Local dump file removed: %s", dumpPath)
	}

	return result, nil
}

// RestorePostgresBackup восстанавливает базу данных из резервной копии
func RestorePostgresBackup(task config.PostgresTask, conn config.PostgresConnection, storagePath string, targetDatabase string, log *logger.Logger) (*PostgresBackupResult, error) {
	result := &PostgresBackupResult{}

	// Расшифровываем пароль
	password := conn.Password

	// Определяем целевую БД
	restoreDB := targetDatabase
	if restoreDB == "" {
		restoreDB = task.Database
	}

	// Создаем временную директорию
	tempDir := "/tmp/postgres_backups"
	if err := os.MkdirAll(tempDir, 0755); err != nil {
		result.Error = fmt.Sprintf("Failed to create temp directory: %v", err)
		return result, fmt.Errorf(result.Error)
	}

	// Скачиваем файл из хранилища
	// Извлекаем имя файла из пути хранилища
	filename := filepath.Base(storagePath)
	if filename == "" || filename == "." {
		// Если не удалось извлечь имя, генерируем временное
		filename = fmt.Sprintf("restore_%d.dump", time.Now().Unix())
	}

	localFile := filepath.Join(tempDir, filename)

	// TODO: Реализовать скачивание из хранилища
	// Пока предполагаем, что файл уже доступен локально или нужно реализовать StorageManager
	log.Warnf("Storage download not yet implemented, assuming file is at: %s", storagePath)

	// Определяем формат по расширению
	fileExt := strings.ToLower(filepath.Ext(localFile))
	var restoreCmd string
	if fileExt == ".dump" || fileExt == ".custom" || fileExt == ".tar" {
		restoreCmd = "pg_restore"
	} else {
		restoreCmd = "psql"
	}

	// Формируем команду восстановления
	var cmd *exec.Cmd
	if restoreCmd == "pg_restore" {
		cmd = exec.Command(
			"pg_restore",
			fmt.Sprintf("--host=%s", conn.Host),
			fmt.Sprintf("--port=%d", conn.Port),
			fmt.Sprintf("--username=%s", conn.Username),
			fmt.Sprintf("--dbname=%s", restoreDB),
			"--clean",
			"--if-exists",
			localFile,
		)
	} else {
		// Для plain SQL используем psql
		cmd = exec.Command(
			"psql",
			fmt.Sprintf("--host=%s", conn.Host),
			fmt.Sprintf("--port=%d", conn.Port),
			fmt.Sprintf("--username=%s", conn.Username),
			fmt.Sprintf("--dbname=%s", restoreDB),
			"--file", localFile,
		)
	}

	cmd.Env = append(os.Environ(), fmt.Sprintf("PGPASSWORD=%s", password))

	log.Infof("Restoring database %s from %s", restoreDB, localFile)

	// Выполняем восстановление
	output, err := cmd.CombinedOutput()
	if err != nil {
		errorMsg := string(output)
		if errorMsg == "" {
			errorMsg = err.Error()
		}
		log.Errorf("Restore failed: %s", errorMsg)
		result.Error = errorMsg
		return result, fmt.Errorf("restore failed: %s", errorMsg)
	}

	result.Success = true
	log.Infof("Database %s restored successfully", restoreDB)

	// Удаляем локальный файл
	if err := os.Remove(localFile); err != nil {
		log.Warnf("Failed to remove local restore file: %v", err)
	}

	return result, nil
}

// uploadPostgresBackupToStorage загружает дамп в хранилище
func uploadPostgresBackupToStorage(task config.PostgresTask, dumpPath string, dbNameSafe string, log *logger.Logger) (string, error) {
	// Парсим конфигурацию хранилища из JSON
	var storageConfig map[string]interface{}
	if err := json.Unmarshal([]byte(task.StorageConfig), &storageConfig); err != nil {
		return "", fmt.Errorf("failed to parse storage config: %v", err)
	}

	storageType := task.StorageType
	if storageType == "" {
		storageType = "s3" // По умолчанию S3
	}

	// Формируем путь в хранилище
	remotePath := fmt.Sprintf("postgres_backups/%s/%s", dbNameSafe, filepath.Base(dumpPath))

	switch storageType {
	case "s3":
		return uploadPostgresBackupToS3(storageConfig, dumpPath, remotePath, log)
	case "local":
		return uploadPostgresBackupToLocalAgent(storageConfig, dumpPath, remotePath, log)
	default:
		return "", fmt.Errorf("storage type %s not yet implemented in agent", storageType)
	}
}

// uploadPostgresBackupToS3 загружает дамп в S3 хранилище
func uploadPostgresBackupToS3(storageConfig map[string]interface{}, dumpPath string, remotePath string, log *logger.Logger) (string, error) {
	// Извлекаем параметры S3 из конфигурации
	endpoint, ok := storageConfig["endpoint"].(string)
	if !ok {
		return "", fmt.Errorf("endpoint not found in storage config")
	}

	accessKey, ok := storageConfig["access_key"].(string)
	if !ok {
		return "", fmt.Errorf("access_key not found in storage config")
	}

	secretKey, ok := storageConfig["secret_key"].(string)
	if !ok {
		return "", fmt.Errorf("secret_key not found in storage config")
	}

	bucketName, ok := storageConfig["bucket_name"].(string)
	if !ok {
		return "", fmt.Errorf("bucket_name not found in storage config")
	}

	region, _ := storageConfig["region"].(string)
	if region == "" {
		region = "us-east-1"
	}

	useSSL := false
	if ssl, ok := storageConfig["use_ssl"].(bool); ok {
		useSSL = ssl
	} else if strings.HasPrefix(endpoint, "https://") {
		useSSL = true
	}

	log.Infof("Uploading PostgreSQL backup to S3: %s/%s", endpoint, bucketName)

	// Очищаем endpoint от протокола
	endpoint = strings.TrimPrefix(strings.TrimPrefix(endpoint, "http://"), "https://")

	// Создаем клиент MinIO
	minioClient, err := minio.New(endpoint, &minio.Options{
		Creds:  credentials.NewStaticV4(accessKey, secretKey, ""),
		Secure: useSSL,
		Region: region,
	})
	if err != nil {
		return "", fmt.Errorf("failed to create S3 client: %v", err)
	}

	ctx := context.Background()

	// Проверяем существование bucket
	exists, err := minioClient.BucketExists(ctx, bucketName)
	if err != nil {
		return "", fmt.Errorf("failed to check bucket: %v", err)
	}

	if !exists {
		if err := minioClient.MakeBucket(ctx, bucketName, minio.MakeBucketOptions{Region: region}); err != nil {
			return "", fmt.Errorf("failed to create bucket: %v", err)
		}
	}

	// Загружаем файл
	_, err = minioClient.FPutObject(ctx, bucketName, remotePath, dumpPath, minio.PutObjectOptions{})
	if err != nil {
		return "", fmt.Errorf("failed to upload file: %v", err)
	}

	s3Path := fmt.Sprintf("s3://%s/%s", bucketName, remotePath)
	log.Infof("Successfully uploaded PostgreSQL backup to %s", s3Path)

	return s3Path, nil
}

// uploadPostgresBackupToLocalAgent загружает дамп на другой агент
func uploadPostgresBackupToLocalAgent(storageConfig map[string]interface{}, dumpPath string, remotePath string, log *logger.Logger) (string, error) {
	agentIP, ok := storageConfig["agent_ip"].(string)
	if !ok {
		return "", fmt.Errorf("agent_ip not found in storage config")
	}

	agentPort, ok := storageConfig["agent_port"].(float64)
	if !ok {
		return "", fmt.Errorf("agent_port not found in storage config")
	}

	basePath, ok := storageConfig["base_path"].(string)
	if !ok {
		return "", fmt.Errorf("base_path not found in storage config")
	}

	// Формируем полный путь на агенте
	fullPath := filepath.Join(basePath, remotePath)
	agentURL := fmt.Sprintf("http://%s:%d/api/storage/upload?base_path=%s&remote_path=%s",
		agentIP, int(agentPort), url.QueryEscape(basePath), url.QueryEscape(remotePath))

	log.Infof("Uploading PostgreSQL backup to local agent: %s", agentURL)

	// Открываем файл
	file, err := os.Open(dumpPath)
	if err != nil {
		return "", fmt.Errorf("failed to open dump file: %v", err)
	}
	defer file.Close()

	// Создаем multipart form
	var b bytes.Buffer
	writer := multipart.NewWriter(&b)
	part, err := writer.CreateFormFile("file", filepath.Base(dumpPath))
	if err != nil {
		return "", fmt.Errorf("failed to create form file: %v", err)
	}

	if _, err := io.Copy(part, file); err != nil {
		return "", fmt.Errorf("failed to copy file: %v", err)
	}
	writer.Close()

	// Отправляем запрос
	resp, err := http.Post(agentURL, writer.FormDataContentType(), &b)
	if err != nil {
		return "", fmt.Errorf("failed to upload file: %v", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(resp.Body)
		return "", fmt.Errorf("upload failed with status %d: %s", resp.StatusCode, string(body))
	}

	var result map[string]interface{}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return "", fmt.Errorf("failed to decode response: %v", err)
	}

	storagePath, ok := result["path"].(string)
	if !ok {
		return "", fmt.Errorf("path not found in response")
	}

	log.Infof("Successfully uploaded PostgreSQL backup to local agent: %s", storagePath)
	return storagePath, nil
}



